bam-index/src/main/java/org/victorchang/QnameSearcher.java

```diff
@@ -29,83 +33,98 @@ public QnameSearcher(KeyPointerReader keyPointerReader, BamRecordReader recordRe
+                    .map(Ascii7Coder.INSTANCE::encode)
+                    .map(input -> new KeyPointer(0, input, input.length))
+                    .collect(toList());
+
+            final Map<byte[], KeyPointer> keyPointerMap = keyPointers.stream()
+                    .collect(toMap(KeyPointer::getKey, Function.identity()));
+
+            keyPointerReader.read(inputStreamLevel1)
+                    .forEach(inputKeyPointer -> {
+                        for (KeyPointer keyPointer : keyPointers) {
+                            if (inputKeyPointer.compareTo(keyPointer) < 0) {
+                                keyPointerMap.put(keyPointer.getKey(), inputKeyPointer);
+                            }
+                        }
+                    });
+            return keyPointerMap;
         }
+    }
 
-        @Override
-        public void onAlignmentPosition(long blockPos, int offset) {
-        }
+    private List<Long> getPointers(List<byte[]> qnames, Path indexFolder, long keyPointer) {
+        try {
+            Path pathLevel0 = indexFolder.resolve("qname.0");
+            FileChannel channelLevel0 = FileChannel.open(pathLevel0, READ);
 
-        @Override
-        public void onQname(byte[] qnameBuffer, int qnameLen) {
-            String decoded = Ascii7Coder.INSTANCE.decode(qnameBuffer, 0, qnameLen);
-            log.info("qname " + decoded);
-        }
+            long compressedOffset = PointerPacker.INSTANCE.unpackCompressedOffset(keyPointer);
+            int unCompressedOffset = PointerPacker.INSTANCE.unpackUnCompressedOffset(keyPointer);
 
-        @Override
-        public void onSequence(byte[] seqBuffer, int seqLen) {
-            String decoded = SeqDecoder.INSTANCE.decode(seqBuffer, 0, seqLen);
-            log.info("sequence " + decoded);
-        }
+            if (compressedOffset >= channelLevel0.size()) {
+                return emptyList();
+            }
 
-        @Override
-        public void onAlignmentRecord(SAMRecord record) {
-            final ByteArrayOutputStream inMemoryStream = new ByteArrayOutputStream();
-            final SAMTextWriter writer = new SAMTextWriter(inMemoryStream);
-            writer.writeAlignment(record);
-            writer.finish();
-            log.debug("SAM alignment record (below)");
-            log.debug(new String(inMemoryStream.toByteArray()));
+            channelLevel0.position(compressedOffset);
+            try (InputStream inputStream = Channels.newInputStream(channelLevel0)) {
+                Iterable<KeyPointer> indexLevel0 = () -> keyPointerReader.read(inputStream, unCompressedOffset).iterator();
+
+                List<byte[]> remainingQnames = new LinkedList<>(qnames);
+                List<Long> pointers = new ArrayList<>();
+                for (KeyPointer x : indexLevel0) {
+                    boolean keysLeft = false;
+                    for (Iterator<byte[]> it = remainingQnames.iterator(); it.hasNext(); ) {
```

Comment:
Hm we could do something more clever here I think. We can keep the `remainingQnames` sorted too. Then we only have to compare the smallest one. If the smallest one doesn't match, the larger ones won't match either, so no point checking them at every step.

Reply:
I thought about that but I didn't do it to keep the code simpler. Looks like it does not make the code hard to read (maybe this way is clearer) so made the change!